### ISSUE TITLE
Add ability to compare functions

### ISSUE THREAD

**Issue by probablyup** [FormidableLabs/react-fast-compare#114]
Perhaps using `Function.toString()`, just need to verify that two functions are the same through non-reference equality means since a lot of people write inline anonymous functions these days

**Comment by ryan-roemer** [comment]
Hi! Thanks for opening the issue. Comparing JS functions by string is a bit risky for reasons of closure/scopes. E.g., a contrived, but simplified example:

```js
let x;
{
  const val = "hi";
  x = () => val;
}

let y;
{
  const val = "ho";
  y = () => val;
}

console.log("===", x === y);
console.log("=== called", x() === y());
console.log("toString", x.toString() === x.toString());
```

I'm also looking at the other "classic deep equals" libraries like lodash `isEqual()` and [`dequal`](https://github.com/lukeed/dequal/issues/1) which don't do this or explicitly chose not to.

Thoughts?

**Comment by probablyup** [comment]
My PoV on this is if the same inline function is composed twice they should come out as equivalent. If there’s any change to the content of the function at all it shouldn’t be considered equal

**Comment by probablyup** [comment]
Actually I’d argue they are the same because they’d reference the same scope variable

**Comment by ryan-roemer** [comment]
Yeah, but they're different scopes (which can happen in like different objects, etc. that are broadly different) and most importantly _produce different rendered results_ which is that a fast equality check is meant to protect you against (a `true` allows you to short-circuit re-rendering).

Put another way, in my contrived example `x` is most definitely not equal to `y` for any programmatic usage :)

**Comment by probablyup** [comment]
We need some sort of handling for functions though, otherwise this library is kinda useless for JSX with inline arrow functions (likely to be a common thing.) I know it’s not perfect but it should cover most common cases ☹️

**Comment by ryan-roemer** [comment]
Hmmm... I do wonder if maybe there's some option of like "allow unsafe function string comparison" or even a custom comparator fn we could take to allow end users to control their own destiny (enable the functionality and then police their own code for safe usage...)

**Comment by exogen** [comment]
I would advise against this as it's unsafe by default as @ryan-roemer suggested. It wouldn't be ideal for the library to be broken-by-default, but might be okay to allow people to opt into unsafe behavior. Here's a more realistic React example demonstrating this:

```jsx
import { useCallback, useState } from "react";

export default function Counter() {
  const [counter, setCounter] = useState(0);

  return (
    <>
      <Button onClick={() => setCounter((value) => value + 1)}>+</button>
      <Button onClick={() => setCounter((value) => value - 1)}>-</button>
      <Button onClick={() => alert(`Value: ${counter}`)}>Get value</button>
    </>
  );
}
```

(Pretend `<Button>` is a component we've memoized using `react-fast-compare` and thus its `onClick` prop would be subject to this function comparison we're talking about.)

In the case of the `+` and `-` buttons, the suggested optimization is totally safe and would work fine: there's no way for those functions to behave differently on each render. But the same optimization would break the `Get value` button: the literal source code of the function doesn't change on each render, but the returned value does (it encloses the `counter` variable). So, clicking it would always result in the alert `Value: 0` even if the value were incremented or decremented.

In theory you could parse the function to see whether there are any non-local variable references within it (references like `window`, `document`, anything global, etc. would also need to be forbidden), but then you have the task of parsing a JavaScript function, and the "fast" part of "react-fast-compare" goes out the window (not to mention the bundle size implications).

In addition to the non-local reference issue, this would likely also break anyone using higher-order functions, as even if you changed a function parameter to another one that has a different `toString()` representation, the function returned by the HOF would *not* change its `toString()`, so you'd again have a bug.

e.g.

```js
const handleClick = createClickHandler(flag ? doThis : doThat);
```

If you imagine what the `createClickHandler` might look like in this example, it doesn't include the actual source of `doThis` or `doThat` within it, it only treats them as a parameter, so its `toString` will always look something like:

```
function createClickHandler(callback) {
  return (...args) => callback(...args);
}
```

Thus even if `doThis` and `doThat` have different `toString()` values, the resulting `handleClick` would not, even if `flag` changes between renders.

If we do add an option to opt into this, I'd recommend making the danger very clear in documentation and naming, prefixing it with `dangerous` or `unsafe`, e.g. `unsafeFunctionalStringEquality: true`.
